### PR TITLE
ttest post

### PR DESCRIPTION
**Pull request by DayalStrub**

### CODE REVIEW

File: posts/bayesian-model-comparison/index.qmd [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+title: Bayesian comparison of cross-validated algorithms
+subtitle: 'Model selection via a Bayesian correlated t-test.'
+date: 2022/11/11
+image: thumb.png
+categories:
+  - ML
+  - Theory
+  - Model evaluation
+  - Bayes
+  - Bayesian estimation
```
Comment (DayalStrub) [inline posts/bayesian-model-comparison/index.qmd:11]: I would just keep Bayes, or are we thinking the more the merrier?

File: posts/bayesian-model-comparison/index.qmd [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+Once we have decided to say that the model with highest F1 score is better, the more serious question arises: how do we know whether a model is indeed better? 
+Can we just take an average score across a test set and compare the numbers, i.e. plain-old pointwise comparison? 
+I think many will get a sense of discomfort just uttering that. 
+We would like something more robust and which, ideally, takes uncertainty into account. 
+
+We ideally want to have a probability distribution which informs us, in some way, what the probability of one model being better than the other is. 
+That is, we want something like $P(M_1 > M_2 | D)$, where $M_1 > M_2$ denotes that model $M_1$ is better than model $M_2$ and $D$ is the data used for the comparison.
+In fact, we may want to do more than that. 
+Ideally, we would want a distribution specifying what the probability of one model being better than another is for each given performance difference. 
+That is, we would like a distribution over the true performance difference of the two algorithms.
```
Comment (DayalStrub) [inline posts/bayesian-model-comparison/index.qmd:55]: This still confuses me a little. Sorry being very slow. How about, simply:

> Ideally, we would want a distribution over the true performance difference of the two algorithms.

File: posts/bayesian-model-comparison/index.qmd [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+What a great motive to enter the wonderful world of *Bayesian modelling*[^1]. 
+
+[^1]: See, e.g. @kruske2013best, for a nice, rapid tour of Bayesian modelling for group comparisons.
+
+## Problem statement
+
+Assume that we are interested in comparing the performance (accuracy, recall, precision, $\text{F}_{\beta}$, etc.) of two predictors.
+If using cross-validation, @bouckaert2003choosing recommends making this more robust (i.e. return more stable statistics) by using *repeated cross-validation*; that is performing evaluation via $m$ repetitions of $k$-fold cross-validation. 
+Both classifiers will be trained on the same data and tested on the same data and we can therefore collect the *differences of their performances*
```
Comment (DayalStrub) [inline posts/bayesian-model-comparison/index.qmd:64]: > Both classifiers will be ~trained on the same data and~ tested on the same data and we can therefore collect the *differences of their performances*

Feel like training is a distraction.

File: posts/bayesian-model-comparison/index.qmd (hunk 1 above)
Comment (gbben) [reply posts/bayesian-model-comparison/index.qmd:11]: Definitely not more the merrier. I wasn't sure and these admittedly are mostly randomise placeholders. "Bayes" however makes me itch a little as we're not really talking about the man himself. "Bayesian"?

File: posts/bayesian-model-comparison/index.qmd (hunk 2 above)
Comment (gbben) [reply posts/bayesian-model-comparison/index.qmd:55]: Yes, much better actually. I'll change

- [ ] change description

File: posts/bayesian-model-comparison/index.qmd (hunk 3 above)
Comment (gbben) [reply posts/bayesian-model-comparison/index.qmd:64]: Hm, I'm not sure. It's important that they also be trained on the same sets as otherwise the models could be generated from different datasets. I could put it in parenthesis?

File: posts/bayesian-model-comparison/index.qmd [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+ax.axvspan(*hdi, alpha=0.2, color='gray', label="94% HDI")
+ax.annotate(f" Mean = {round(mean, 3)}", xy=(mean, y_mean + 0.5))
+ax.annotate(f" P(svm > rf) = {round(prob_svm_better, 4)}", xy=(-x_lim, y_mean/2))
+ax.annotate(f" P(svm ~ rf) \n         = \n      {round(prob_rope, 4)}", xy=(ROPE[0], y_mean/2))
+ax.annotate(f" P(svm < rf) = {round(prob_rf_better, 4)}", xy=(mean + 0.75*(hdi[1]-mean), y_mean/2))
+
+ax.legend();
+```
+
+Since the posterior distribution informs us about the relative credibility values across the reals, from the posterior we get the uncertainty in the estimate.
```
Comment (DayalStrub) [inline posts/bayesian-model-comparison/index.qmd:394]: can we review this paragraph and bullets together?

File: posts/bayesian-model-comparison/index.qmd [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+
+1) estimate the posterior probability of a reasonable null hypothesis, i.e. if the difference in performance is within a couple of percentage points they may well be considered equivalent. 
+This will be given by the area within the rope region, above denoted by $P(M_1 \sim M_2)$.
+2) significant differences also have practical meaning (area outside the ROPE)
+3) represents effect size and uncertainty (HDI)
+4) we can make meaningful decision based on requiring that the probability of a model being better than another has to be at least 95%, say.
+
+## Cost sensitive decisions
+We will be dealing with this in more detail in a separate blog post. 
+In the real world, it is usually the case that we want to reason about situations and make decisions based on costs and errors.
```
Comment (DayalStrub) [inline posts/bayesian-model-comparison/index.qmd:406]: "and errors"?

File: posts/bayesian-model-comparison/index.qmd [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+Phrased differently, the choice of the cost measure should depend on how the system is going to be used, rather than on any inherent specification of the training process.
+The issue with doing this is that it is hard. 
+It is hard because specifying a cost-aware loss function is non-trivial, because cost-specifications are domain specific, and becuase even in the case of roughly knowing what the costs are, using this information is hard, i.e. the specified weigthed cost may be a difficult objective for optimisers to work with.
+However if we can specify costs then decision making based on these would be the best way to work as this allows one to take into consideration the utility of the decision that will be made.
+
+In our situation, in order to do cost-sensitive analysis/decision making, all we need is to specify a cost function -- we won't need to run an optimisation on this objective function. 
+This is a function that defines the loss (or cost) we incur in making a given decision (e.g. the wrong decision).
+A typical example is whether to give more importance to a false positive or a false negative.
```
Comment (DayalStrub) [inline posts/bayesian-model-comparison/index.qmd:414]: wouldn't this be picked up by error/score, ie F1 or otherwise?

File: posts/bayesian-model-comparison/index.qmd [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+expected_cost = cost_matrix.dot(probabilities)
+```
+```{python}
+#| echo: false
+print(f"\nCost of deciding on SVM:", round(expected_cost[0], 4))
+print(f"\nCost of deciding on RF:", round(expected_cost[1], 4))
+```
+
+The lowest cost would determine the optimal decision.
+We see that we would incur a significant cost in choosing the SVM over the RF.
```
Comment (DayalStrub) [inline posts/bayesian-model-comparison/index.qmd:468]: i wonder whether it would have been interesting to think up a hypothethical in which the difference is reversed, or not as pronounced when taking into account the cost, despite the probability being so skewed.

File: posts/bayesian-model-comparison/index.qmd [hunk 8]
```diff
@@ -0,0 +1,470 @@
...
+  + -5 if they are equivalent as we save compute time, 
+  + 2 if RF is better as we would lose performance
+
+* cost of choosing RF is:
+  + 7 if SVM is better because we pay for computational and performance cost, 
+  + 5 if they are equivalent as we add to the compute time, 
+  + 0 if RF is indeed better
+
+
+The **expected cost** can then be obtained by multiplying the cost_matrix with the posterior probabilities
```
Comment (DayalStrub) [inline posts/bayesian-model-comparison/index.qmd:450]: i find it odd to think of/refer to the cumulative probabilities as posterior, given we dont set priors and we dont compute them as part of the modelling

File: posts/bayesian-model-comparison/index.qmd (hunk 6 above)
Comment (gbben) [reply posts/bayesian-model-comparison/index.qmd:414]: In some sense, yes. However without specifying exactly what a false negative or false positive costs you you can't really define a good beta. And even if you do that harmonic mean of precision and recall doesn't necessarily become granular enough. Fbeta does go towards that though, yes. 

Maybe I should expand on this/clarify

File: posts/bayesian-model-comparison/index.qmd (hunk 3 above)
Comment (DayalStrub) [reply posts/bayesian-model-comparison/index.qmd:63]: interesting. would this be a problem for computing delta?

i see below training and testing become intertwined due to cross-validation, hence rho. but not sure i see it here.

File: posts/bayesian-model-comparison/index.qmd (hunk 7 above)
Comment (gbben) [reply posts/bayesian-model-comparison/index.qmd:468]: That's certainly possible. However in this case the models are so similar in latency and memoryy etc that it's hard to justify. If RF had been neural then I think one could concoct costs so extreme that you would get what you want. 

I think the options for something more interesting are:
1) chose models for which distribution isn't so much towards the right
2) using a larger rope
3) using "graded" partitions for the cost. Something I'd mentioned to you and that I'd like to show you

File: posts/bayesian-model-comparison/index.qmd (hunk 6 above)
Comment (DayalStrub) [reply posts/bayesian-model-comparison/index.qmd:414]: nah. i fear we will be expanding on everything. i get you point. happy to leave as is and resolve conversation

File: posts/bayesian-model-comparison/index.qmd (hunk 8 above)
Comment (gbben) [reply posts/bayesian-model-comparison/index.qmd]: The t-dist here is a posterior though, so we are getting the probabilities given by volumes in the posterior. Am I misunderstanding?

File: posts/bayesian-model-comparison/index.qmd (hunk 3 above)
Comment (gbben) [reply posts/bayesian-model-comparison/index.qmd:63]: The rho correlation is the correction because of the samples not being independent (Samples being the scores, or difference of scores also) however we also need to have the same train set and same test set for the battle to be fair. Even in the case of plain train-test split it would be unfair to have one more trained on partition A and tested on C and another trained on B and tested on C.

File: posts/bayesian-model-comparison/index.qmd (hunk 5 above)
Comment (gbben) [reply posts/bayesian-model-comparison/index.qmd:406]: Too much? Too much.

File: posts/bayesian-model-comparison/index.qmd (hunk 7 above)
Comment (DayalStrub) [reply posts/bayesian-model-comparison/index.qmd:468]: i think we are both on the same page, so i say we drop my comment and leave as is, as 1 and 2 are much hassle for little gain, and 3 is very cool but we should do in the next post.

File: posts/bayesian-model-comparison/index.qmd (hunk 8 above)
Comment (DayalStrub) [reply posts/bayesian-model-comparison/index.qmd]: what is a posterior? 

i so wanted to just comment w that! i just mean that for this distribution (over {M1<M2, M1=M2, M1>M2}) we have no prior. it feels like a "derived" distribution from the (main) posterior. but i get your point, and happy to leave!

File: posts/bayesian-model-comparison/index.qmd (hunk 3 above)
Comment (DayalStrub) [reply posts/bayesian-model-comparison/index.qmd:63]: agree about like-for-like, not sure it's necessary tho. and indeed we dont say so.

think we agree, so happy to leave.

File: posts/bayesian-model-comparison/index.qmd (hunk 5 above)
Comment (DayalStrub) [reply posts/bayesian-model-comparison/index.qmd:406]: Que pasa?

File: posts/bayesian-model-comparison/index.qmd (hunk 8 above)
Comment (gbben) [reply posts/bayesian-model-comparison/index.qmd]: Aha, no yes, you are right. We derive this, that's true. I guess you could still call this P(M1 > M2 |D) and not talk about how you would get this directly. After all we could do the same by getting P(M1>M2) based on the Normal prior. Then we perform inference, get a t-dist, then derive the probability.

File: posts/bayesian-model-comparison/index.qmd (hunk 2 above)
Comment (gbben) [reply posts/bayesian-model-comparison/index.qmd:54]: Ha, just realised I'd written exactly that below. had misunderstood your comment. Will delete the "duplication"

File: posts/bayesian-model-comparison/index.qmd (hunk 3 above)
Comment (gbben) [reply posts/bayesian-model-comparison/index.qmd:63]: I'll leave it then.

File: posts/bayesian-model-comparison/index.qmd (hunk 8 above)
Comment (gbben) [reply posts/bayesian-model-comparison/index.qmd]: Do you want me to make this clearer?

Comment (DayalStrub) [reply posts/bayesian-model-comparison/index.qmd]: isn't the probability actually P(X|D)  where X \in {M1<M2, M1=M2, M1>M2}, or are we really thinking of as 3 probabilities. i guess above we are.

i think if we have a quick fix could be good. but it's minor. how about 

> The **expected cost** can then be obtained by multiplying the cost_matrix with the probabilities P(M1 > M2 |D), P(M1 ~ M2 |D), and P(M1 < M2 |D)

Comment (gbben) [reply posts/bayesian-model-comparison/index.qmd]: I'll clarify 👍 . I stress though that these are indeed posteriors. Further, they can be considered as cumulative of the actual posterior.  Once you have those then you can reformulate these are parameters of a multinoulli distribution, just like you can take the cumulative of a gaussian and feed it as parameter of a Bernoulli distribution.

**Review by gbben — APPROVED** [review]
